chapters/untyped_pointers.md

```diff
@@ -0,0 +1,129 @@
+
+With untyped pointer, we still want will use the access chain to get the `StorageClass`, but we can now remove the redundancy by first turning the pointer into an untyped pointer.
+
+```diff
+   %uint = OpTypeInt 32 0
+ %uint_0 = OpConstant %uint 0
+
+- %ptr = OpTypePointer StorageBuffer %uint
++ %ptr = OpTypeUntypedPointerKHR StorageBuffer
+
+    %SSBO = OpTypeStruct %uint
+%SSBO_ptr = OpTypePointer StorageBuffer %SSBO
+%variable = OpVariable %SSBO_ptr StorageBuffer
+
+  %ac = OpAccessChain %ptr %variable %uint_0
+%load = OpLoad %uint %ac
+        OpStore %ac %uint_0
+```
+
+But this is **stil invalid**! We now need a matching untyped access chain as well where we now have a new `Base Type` operand.
+
+```diff
+   %uint = OpTypeInt 32 0
+ %uint_0 = OpConstant %uint 0
+
+ %ptr = OpTypeUntypedPointerKHR StorageBuffer
+
+    %SSBO = OpTypeStruct %uint
+%SSBO_ptr = OpTypePointer StorageBuffer %SSBO
+%variable = OpVariable %SSBO_ptr StorageBuffer
+
+-  %ac = OpAccessChain %ptr %variable %uint_0
++  %ac = OpUntypedAccessChainKHR %ptr %SSBO %variable %uint_0
+%load = OpLoad %uint %ac
+        OpStore %ac %uint_0
+```
+
+From here, we can now go an extra step and turn the `OpVariable` into an untyped variable as it now doesn't requires to be tied to a pointer anymore.
+
+This is the same mental model of taking a `ssbo* ptr` and now making it `void* ptr`
+
+```diff
+   %uint = OpTypeInt 32 0
+ %uint_0 = OpConstant %uint 0
+
+ %ptr = OpTypeUntypedPointerKHR StorageBuffer
+
+    %SSBO = OpTypeStruct %uint
+- %SSBO_ptr = OpTypePointer StorageBuffer %SSBO
+- %variable = OpVariable %SSBO_ptr StorageBuffer
++ %variable = OpUntypedVariableKHR %ptr StorageBuffer %SSBO
+
+  %ac = OpUntypedAccessChainKHR %ptr %SSBO %variable %uint_0
+%load = OpLoad %uint %ac
+        OpStore %ac %uint_0
+```
+
+And with that, we are done and have now turned the original SPIR-V to using untyped pointers and still have the same semantic meaning.
+
+> Final SPIR-V https://godbolt.org/z/1n9bffq4d
+
+# Removing accidental assumptions
+
+What people parsing SPIR-V need to care about is where, by accident, they are getting the type from the `OpTypePointer`.
+
+For `OpLoad`, use the `Result Type` operand instead of the `Pointer` operand.
+
+For `OpStore`, use the `Object` operand instead of the `Pointer` operand.
+
+For `OpCopyMemory`, either the `Source` or `Target` operand will contain a `OpTypePointer` still.
+
+For `OpCopyMemorSized`, both the `Source` and `Target` could be an untyped pointer, in this case, it interprets the data as an array of 8-bit integers.
+
+For `OpVariable`, there is an optional `Data Type` in the `OpUntypedVariableKHR` to check. If you are using `Shader` (Vulkan) then this is [required the be there](https://docs.vulkan.org/spec/latest/appendices/spirvenv.html#VUID-StandaloneSpirv-OpUntypedVariableKHR-11167).
```

Comment:
For the `Shader` bit, Vulkan requires that the storage class be explicitly laid out, so the following:
* StorageBuffer
* PhysicalStorageBuffer
* Uniform
* PushConstant
* Workgroup (with workgroup memory explicit layout extension)